exp/bufferpool/pool.go

```diff
@@ -4,49 +4,161 @@ package bufferpool
 import "sync"
 
 const (
-	nBuckets = 20
+	defaultMinSize     = 1024
+	defaultBucketCount = 10
 )
 
-// A default global pool.
-var Default Pool
-
-// A pool of buffers, in variable sizes.
-type Pool struct {
-	buckets [nBuckets]sync.Pool
-}
-
-// Get a buffer of the given length. Its capacity may be larger than the
-// requested size.
-func (p *Pool) Get(size int) []byte {
-	for i := 0; i < nBuckets; i++ {
-		capacity := 1 << i
-		if capacity >= size {
-			var ret []byte
-			item := p.buckets[i].Get()
-			if item == nil {
-				ret = make([]byte, capacity)
-			} else {
-				ret = item.([]byte)
-			}
-			ret = ret[:size]
-			return ret
-		}
+var pool Pool = &BasicPool{}
+
+// Set the global pool.  This can be used to override the default implementation
+// to optimize for different allocation profiles. For example, applications that
+// target embedded platforms may wish to lower the minimum buffer size from 1KiB
+// to 512B, or less.
+//
+// The default implementation uses BasicPool, which is suitable for a wide range
+// of applications.  Most applications will not benefit from using a custom pool,
+// so be sure to profile.
+//
+// Set panics if p == nil.
+func Set(p Pool) {
+	if p == nil {
+		panic("Pool(nil)")
 	}
-	return make([]byte, size)
+
+	pool = p
+}
+
+// Get a buffer of with the specified size from the global pool.  The
+// resulting buffer's capacity MAY exceed size. The buffer SHOULD NOT
+// be resized beyond its underlying capacity, or it will leak to the
+// garbage collector.
```

Comment:
I might just replace all this with something to the effect of `calls .Get() on the global Pool`, maybe moving some of the details here to that doc comment. Same thing applies to Put()

Reply:
Done